### PR TITLE
Feature/clean up projection classes for deprecation plan

### PR DESCRIPTION
**Pull request by MiguelRicardos**

### CODE REVIEW

**Review by cogross — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java [hunk 1]
```diff
@@ -38,40 +38,43 @@ public class DocumentProjection implements DocumentPermutation {
     private boolean trackSizes = true;
 
     public DocumentProjection() {
-        this(false, false);
+        this.includeGroupingContext = false;
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java]: Why not leave this as : this(false, false)

Comment (MiguelRicardos) [reply warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java]: I get a compilation failure when I do it that way.

**Review by apmoriarty — APPROVED** [review]

File: warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java]: You can change this constructor to
```java
public DocumentProjection() {
    this(false, false);
}
```
without getting a compilation error

File: warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java [hunk 2]
```diff
@@ -38,40 +38,43 @@ public class DocumentProjection implements DocumentPermutation {
...
 
     public DocumentProjection(boolean includeGroupingContext, boolean reducedResponse) {
-        this(includeGroupingContext, reducedResponse, true);
+        this.includeGroupingContext = includeGroupingContext;
+        this.reducedResponse = reducedResponse;
+        this.projection = new Projection();
     }
 
-    public DocumentProjection(boolean includeGroupingContext, boolean reducedResponse, boolean trackSizes) {
+    public DocumentProjection(boolean includeGroupingContext, boolean reducedResponse, boolean trackSizes, Set<String> projections,
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java:52]: Deprecate the existing constructors.

File: warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java [hunk 3]
```diff
@@ -38,40 +38,43 @@ public class DocumentProjection implements DocumentPermutation {
...
         this.projection.setIncludes(includes);
     }
 
-    /**
-     * Configure the delegate {@link Projection} with the fields to exclude
-     *
-     * @param excludes
-     *            the set of fields to exclude
-     */
+    public DocumentProjection(Set<String> projections, Projection.ProjectionType projectionType) {
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java:64]: please group constructors together at the top of the class.

**Review by apmoriarty — CHANGES_REQUESTED** [review]
None of the changes requested are functional changes, so there's no need to retest.

File: warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java (hunk 2 above)
Comment (MiguelRicardos) [reply warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java:52]: rgr

File: warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java (hunk 3 above)
Comment (MiguelRicardos) [reply warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java:64]: you got it.

File: warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java]: This change wasn't made?

Comment (MiguelRicardos) [reply warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java]: Change has been made.

File: warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java [hunk 4]
```diff
@@ -37,37 +37,42 @@ public class DocumentProjection implements DocumentPermutation {
...
-     * @param includes
-     *            the set of fields to include
-     */
+    @Deprecated
+    public DocumentProjection(Set<String> projections, Projection.ProjectionType projectionType) {
+        this(false, false, projections, projectionType);
+    }
+
+    @Deprecated
+    public DocumentProjection(boolean includeGroupingContext, boolean reducedResponse, Set<String> projections, Projection.ProjectionType projectionType) {
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java:69]: Does the new constructor need to be deprecated?

File: warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java [hunk 5]
```diff
@@ -37,37 +37,42 @@ public class DocumentProjection implements DocumentPermutation {
...
     }
 
-    /**
-     * Set the delegate {@link Projection} with the fields to include
-     *
-     * @param includes
-     *            the set of fields to include
-     */
+    @Deprecated
+    public DocumentProjection(Set<String> projections, Projection.ProjectionType projectionType) {
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java:64]: Add documentation that this constructor will not include the grouping context and will not reduce the document.

File: warehouse/query-core/src/main/java/datawave/query/iterator/QueryIterator.java [hunk 6]
```diff
@@ -1208,22 +1210,14 @@ private void debugBooleanLogicIterator(NestedIterator<Key> root, String prefix)
...
+                            Projection.ProjectionType.INCLUDES);
         } else if (this.useBlackListedFields) {
-            // make sure we are not excluding any fields being matched in the limit fields mechanism
-            if (!this.matchingFieldSets.isEmpty()) {
-                this.blackListedFields.removeAll(getMatchingFieldList());
-            }
-            projection.setExcludes(this.blackListedFields);
-            return projection;
+            return new DocumentProjection(this.isIncludeGroupingContext(), this.isReducedResponse(), isTrackSizes(), this.blackListedFields,
+                            Projection.ProjectionType.EXCLUDES);
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/iterator/QueryIterator.java]: This change lost the removal of matching field sets. I think a better approach in this case is to build the fields first, then the projection, then the document projection.


And typing this out I realize we actually need to add a constructor for document projection that looks like this
```java
public DocumentProjection(boolean includeGroupingContext, boolean isReducedResponse, boolean isTrackSizes, Projection projection){
   ...
}
```

File: warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java (hunk 4 above)
Comment (MiguelRicardos) [reply warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java:69]: This constructor is from the parent class that will be removed in the future.

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/function/DocumentProjection.java:69]: Perfect, I missed that

File: warehouse/query-core/src/main/java/datawave/query/iterator/QueryIterator.java [hunk 7]
```diff
@@ -1208,22 +1210,14 @@ private void debugBooleanLogicIterator(NestedIterator<Key> root, String prefix)
...
-        DocumentProjection projection = new DocumentProjection(this.isIncludeGroupingContext(), this.isReducedResponse(), isTrackSizes());
 
         if (this.useWhiteListedFields) {
-            // make sure we include any fields being matched in the limit fields mechanism
-            if (!this.matchingFieldSets.isEmpty()) {
-                this.whiteListedFields.addAll(getMatchingFieldList());
-            }
-            projection.setIncludes(this.whiteListedFields);
-            return projection;
+
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/iterator/QueryIterator.java]: Need to add this bit
```java
            // make sure we include any fields being matched in the limit fields mechanism
            if (!this.matchingFieldSets.isEmpty()) {
                this.whiteListedFields.addAll(getMatchingFieldList());
            }
```

File: warehouse/query-core/src/main/java/datawave/query/iterator/QueryIterator.java [hunk 8]
```diff
@@ -1208,22 +1210,14 @@ private void debugBooleanLogicIterator(NestedIterator<Key> root, String prefix)
...
+            return new DocumentProjection(this.isIncludeGroupingContext(), this.isReducedResponse(), isTrackSizes(),
+                            new Projection(this.whiteListedFields, Projection.ProjectionType.INCLUDES));
         } else if (this.useBlackListedFields) {
-            // make sure we are not excluding any fields being matched in the limit fields mechanism
-            if (!this.matchingFieldSets.isEmpty()) {
-                this.blackListedFields.removeAll(getMatchingFieldList());
-            }
-            projection.setExcludes(this.blackListedFields);
-            return projection;
+            return new DocumentProjection(this.isIncludeGroupingContext(), this.isReducedResponse(), isTrackSizes(),
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/iterator/QueryIterator.java:1226]: Also need to handle the exclude fields
```java
            // make sure we are not excluding any fields being matched in the limit fields mechanism
            if (!this.matchingFieldSets.isEmpty()) {
                this.blackListedFields.removeAll(getMatchingFieldList());
            }
```

Comment (MiguelRicardos) [reply warehouse/query-core/src/main/java/datawave/query/iterator/QueryIterator.java:1226]: added this back.

File: warehouse/query-core/src/main/java/datawave/query/iterator/QueryIterator.java (hunk 6 above)
Comment (MiguelRicardos) [reply warehouse/query-core/src/main/java/datawave/query/iterator/QueryIterator.java]: added this back.

**Review by apmoriarty — APPROVED** [review]

**Review by cogross — APPROVED** [review]

**Pull request merged**